### PR TITLE
feat: waitosaur

### PR DESCRIPTION
**Pull request by ratik**
This pull request introduces a new upgradable smart contract, `Waitosaur`, for managing the locking and unlocking of ERC20 tokens with role-based access, along the tests. The contract leverages OpenZeppelin's upgradable patterns and includes error handling and event emission. The test suite uses a mock ERC20 token and covers core functionalities and edge cases

### CODE REVIEW

File: src/Waitosaur.sol [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+        require(success && data.length >= 32, "balanceOf failed");
+        uint256 balance = abi.decode(data, (uint256));
+        if (
+            balance <
+            StorageSlot.getUint256Slot(LOCKED_AMOUNT_STORAGE_SLOT).value
+        ) revert InsufficientBalance();
+        if (
+            StorageSlot.getAddressSlot(RECEIVER_STORAGE_SLOT).value ==
+            address(0)
+        ) revert("Receiver not set");
```
Comment (albertandrejev) [inline src/Waitosaur.sol]: I guess it is better to check for this condition in the initializer and in appropriate update method

File: src/Waitosaur.sol [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    ) public initializer {
+        if (_token == address(0)) revert InvalidTokenAddress();
+        __Ownable_init(owner_);
+        __UUPSUpgradeable_init();
+        StorageSlot.getUint256Slot(LOCKED_AMOUNT_STORAGE_SLOT).value = 0;
+        StorageSlot.getBooleanSlot(UNLOCKED_STORAGE_SLOT).value = true;
+        StorageSlot.getAddressSlot(ERC20_STORAGE_SLOT).value = _token;
+        StorageSlot.getAddressSlot(LOCKER_STORAGE_SLOT).value = _locker;
+        StorageSlot.getAddressSlot(UNLOCKER_STORAGE_SLOT).value = _unLocker;
+        StorageSlot.getAddressSlot(RECEIVER_STORAGE_SLOT).value = _receiver;
```
Comment (albertandrejev) [inline src/Waitosaur.sol]: Please check this values for zero address, and in update methods as well

File: test/Waitosaur.test.sol [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+            address(token),
+            locker,
+            unLocker,
+            receiver
+        );
+        // Deploy proxy with initializer
+        ERC1967Proxy proxy = new ERC1967Proxy(address(logic), data);
+        waitosaur = waitosaurSrc.Waitosaur(address(proxy));
+        token.mint(address(waitosaur), 1000 ether);
+    }
```
Comment (albertandrejev) [inline test/Waitosaur.test.sol]: please move this to the top of current test contract for better readability

File: src/Waitosaur.sol [hunk 4]
```diff
@@ -0,0 +1,176 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.25;
```
Comment (foxpy) [inline src/Waitosaur.sol]: We use `^0.8.28` in all other contracts, please keep it consistent.

File: test/Waitosaur.test.sol [hunk 5]
```diff
@@ -0,0 +1,124 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.25;
```
Comment (foxpy) [inline test/Waitosaur.test.sol]: We use `^0.8.28` in all other contracts, please keep it consistent.

File: src/Waitosaur.sol [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+
+    function unLocker() public view returns (address) {
+        return StorageSlot.getAddressSlot(UNLOCKER_STORAGE_SLOT).value;
+    }
+
+    function _authorizeUpgrade(
+        address newImplementation
+    ) internal override onlyOwner {}
+
+    uint256[50] private __gap;
```
Comment (foxpy) [inline src/Waitosaur.sol]: Do we still need it?

File: src/WaitosaurHolder.sol
Comment (foxpy) [inline src/WaitosaurHolder.sol:1]: I see that a lot of storage values are tightly related to each other and used together in different functions. Maybe we should just create a config struct which holds them all in one place?

File: src/Waitosaur.sol [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+            StorageSlot.getAddressSlot(UNLOCKER_STORAGE_SLOT).value
+        ) revert NotUnLocker();
+        if (StorageSlot.getBooleanSlot(UNLOCKED_STORAGE_SLOT).value == true)
+            revert AlreadyUnlocked();
+        if (StorageSlot.getUint256Slot(LOCKED_AMOUNT_STORAGE_SLOT).value == 0)
+            revert NothingLocked();
+        address tokenAddr = token();
+        (bool success, bytes memory data) = tokenAddr.call(
+            abi.encodeWithSignature("balanceOf(address)", address(this))
+        );
```
Comment (foxpy) [inline src/Waitosaur.sol]: you can import `IERC20` interface, convert token `address` to this type and call `balanceOf()` on it directly.

File: src/Waitosaur.sol [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+            abi.encodeWithSignature(
+                "transfer(address,uint256)",
+                StorageSlot.getAddressSlot(RECEIVER_STORAGE_SLOT).value,
+                StorageSlot.getUint256Slot(LOCKED_AMOUNT_STORAGE_SLOT).value
+            )
+        );
+        require(
+            success && (data.length == 0 || abi.decode(data, (bool))),
+            "Transfer failed"
+        );
```
Comment (foxpy) [inline src/Waitosaur.sol]: That looks scary. Please, use `SafeERC20`, like here: https://github.com/structured-org/erc20-converter/blob/main/src/ERC20Converter.sol

**Review by foxpy — CHANGES_REQUESTED** [review]

**Review by albertandrejev — APPROVED** [review]

File: src/Waitosaur.sol (hunk 4 above)
Comment (foxpy) [reply src/Waitosaur.sol]: Please note the `^` symbol, it is important because it prevents newer (e.g. `0.9.0`) Solidity versions from being used.

**Comment by albertandrejev** [comment]
Please change name of the contract to WaitosaurHolder

File: src/WaitosaurHolder.sol [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+    error NotUnLocker();
+
+    /// @dev keccak256(abi.encode(uint256(keccak256("maxbtc.waitosaur.config")) - 1)) & ~bytes32(uint256(0xff))
+    bytes32 private constant CONFIG_STORAGE_SLOT =
+        0x8f3661154de8aca85dd40a286ee737a27a3b740a5d65a00dc9530f3fcf760200;
+    /// @dev keccak256(abi.encode(uint256(keccak256("maxbtc.waitosaur.state")) - 1)) & ~bytes32(uint256(0xff))
+    bytes32 private constant STATE_STORAGE_SLOT =
+        0xcb11509986cbc674883f88160c41f636c73566a871549bf8441ad8bc0e648300;
+
+    event Locked(uint256 indexed amount);
```
Comment (albertandrejev) [inline src/WaitosaurHolder.sol]: I guess `indexed` is not required here

File: src/WaitosaurHolder.sol [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+    event Unlocked();
+    event ConfigUpdated(WaitosaurConfig indexed config);
```
Comment (albertandrejev) [inline src/WaitosaurHolder.sol]: and here as well

File: src/WaitosaurHolder.sol [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+        WaitosaurConfig storage config = _getWaitosaurConfig();
+        config.locker = _locker;
+        config.unLocker = _unLocker;
+        config.receiver = _receiver;
+        config.token = _token;
+    }
+
+    function lock(uint256 amount) external {
+        WaitosaurConfig storage config = _getWaitosaurConfig();
+        if (msg.sender != config.locker) revert NotLocker();
```
Comment (foxpy) [inline src/WaitosaurHolder.sol]: I suggest using `_msgSender()`everywhere:
https://forum.openzeppelin.com/t/why-use-msgsender-rather-than-msg-sender/4370

Comment (foxpy) [reply src/WaitosaurHolder.sol]: And the same in Withdrawal Token... sorry for not catching that earlier!

File: src/WaitosaurHolder.sol [hunk 12]
```diff
@@ -0,0 +1,167 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.28;
```
Comment (foxpy) [inline src/WaitosaurHolder.sol]:
```suggestion
pragma solidity ^0.8.28;
```

File: test/WaitosaurHolder.test.sol [hunk 13]
```diff
@@ -0,0 +1,226 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.28;
```
Comment (foxpy) [inline test/WaitosaurHolder.test.sol]:
```suggestion
pragma solidity ^0.8.28;
```

**Review by albertandrejev — APPROVED** [review]

**Review by foxpy — APPROVED** [review]
